### PR TITLE
Accept all constructable function forms in Reflect.construct

### LINKED ISSUE #516: Engine: Reflect.construct rejects function expressions as proxy target
## Engine bug: `Reflect.construct` rejects `function` declarations and expressions as the proxy target

`Reflect.construct(target, args, newTarget)` is supposed to call `[[Construct]]` of `target` with `newTarget` substituted. Per spec, every function created through `function` syntax (declaration or expression) has a `[[Construct]]` slot — they are constructible. Functions created through other forms (arrow, class methods, object shorthand methods, async, generators, async generators) do NOT have `[[Construct]]` and correctly throw.

Goccia rejects `function` declarations and expressions too — only `class` declarations are accepted as the `Reflect.construct` target.

### Scope (validated; tested in both modes, identical results)

| Function-creation form                | Spec expectation | Goccia actual |
|---|---|---|
| `function fnDecl() {}`                | constructible    | ❌ rejected     |
| `function () {}` (expression)         | constructible    | ❌ rejected     |
| `function named() {}` (named expr)    | constructible    | ❌ rejected     |
| `(...) => {}` (arrow)                 | not constructible | ✅ rejected (correct) |
| `({ m() {} }).m` (object shorthand)   | not constructible | ✅ rejected (correct) |
| `(class { m() {} }).prototype.m`      | not constructible | ✅ rejected (correct) |
| `async function () {}`                | not constructible | ✅ rejected (correct) |
| `function* () {}`                     | not constructible | ✅ rejected (correct) |
| `async function* () {}`               | not constructible | ✅ rejected (correct) |
| `class C {}`                          | constructible    | ✅ accepted (correct) |
| `Int8Array`, `Object`, etc.           | constructible    | ✅ accepted (correct) |
| `class extends Error {}`              | constructible    | ✅ accepted (correct) |
| `(function(){}).bind(null)`           | constructible (inherits from target) | ❌ rejected (because target is rejected) |

### Reproduce

```bash
./build.pas loaderbare
cat > /tmp/probe.js <<'EOJS'
function fnDecl() {}
try { Reflect.construct(fnDecl, [], Int8Array);  print("fn decl: OK"); } catch (e) { print("fn decl: THROWS"); }
try { Reflect.construct(function(){}, [], Int8Array); print("fn expr: OK"); } catch (e) { print("fn expr: THROWS"); }
try { Reflect.construct(class {}, [], Int8Array); print("class:   OK"); } catch (e) { print("class:   THROWS"); }
try { Reflect.construct(() => {}, [], Int8Array); print("arrow:   OK"); } catch (e) { print("arrow:   THROWS"); }  // expected THROWS
EOJS
./build/GocciaScriptLoaderBare --asi --compat-all --unsafe-function-constructor /tmp/probe.js
```
Output:
```
fn decl: THROWS    ← BUG (should be OK)
fn expr: THROWS    ← BUG (should be OK)
class:   OK
arrow:   THROWS    ← correct
```

### Why this matters

Stock test262 `harness/isConstructor.js` uses `Reflect.construct(function(){}, [], f)` to probe whether `f` is constructible. With this bug, every test262 test that uses the `isConstructor` helper (via `includes: [isConstructor.js]`) fails the probe. The current orchestrator works around it by bundling `scripts/test262_harness/isConstructor.js` that uses `Reflect.construct(class {}, [], f)`.

Likely fix area: `Reflect.construct` validates the target via the same predicate that gates the `new` operator. For `function` declarations/expressions, that predicate currently returns "not constructible." Per spec the predicate should be true for all `Function` instances except those whose `[[ConstructorKind]]` is "derived without super-call" or whose internal `[[Construct]]` slot is absent (arrow, methods, async, generators). With `--compat-function` enabled, function declarations and expressions should land in the constructible bucket.

### Surfaced by

#513 (test262 harness rewrite).

### PR DESCRIPTION
**Pull request by frostney**
## Summary

- `Reflect.construct` previously rejected function declarations, function expressions, and bound functions because the local `IsConstructorValue` predicate only recognized classes and constructable native functions, while `new fn()` already accepted them via the value system's virtual `IsConstructable`.
- Defer the predicate to `TGocciaValue.IsConstructable` (the same hook used by the `new` operator in both interpreter and bytecode VM), and expand the construction dispatch to walk bound-function chains, allocate a receiver from `newTarget`'s prototype per ES2026 §10.1.14, and route proxies through `ConstructTrap`.
- Adds gated tests under `tests/built-ins/Reflect/construct/` (with `goccia.json` enabling `compat-function`) for function-keyword targets and bound-function-keyword targets, plus ungated tests in `construct.js` for arrow/method rejection, bound-class construction, proxy traps, and arrow-as-newTarget rejection.

## Test plan

- [x] `./build.pas testrunner && ./build/GocciaTestRunner tests/built-ins/Reflect --asi --no-progress` (interpreter): 133 pass, 0 fail
- [x] Same with `--mode=bytecode`: 133 pass, 0 fail
- [x] `./build/GocciaTestRunner tests --asi --unsafe-ffi --no-progress` (full suite, interpreter): 8628 pass, 0 fail
- [x] Same with `--mode=bytecode`: 8628 pass, 0 fail
- [x] `./format.pas --check`: clean
- [x] Issue's reproduction script (`fn decl`, `fn expr`, `class`, `arrow`) now reports OK/OK/OK/THROWS as the spec table expects.

Closes #516

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: source/units/Goccia.Builtins.GlobalReflect.pas [hunk 1]
```diff
@@ -129,25 +131,83 @@ function TGocciaGlobalReflect.ReflectApply(const AArgs: TGocciaArgumentsCollecti
...
+
+  if ProtoValue is TGocciaObjectValue then
+    Result := TGocciaObjectValue(ProtoValue)
+  else
+  begin
+    if not Assigned(TGocciaObjectValue.SharedObjectPrototype) then
+      TGocciaObjectValue.InitializeSharedPrototype;
+    Result := TGocciaObjectValue.SharedObjectPrototype;
+  end;
+end;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Builtins.GlobalReflect.pas]: Good catch — fixed in d47f960f.

Dropped `UnwrapBoundFunction` from `GetProtoFromConstructor` so it now does `Get(newTarget, "prototype")` directly per §10.1.14, with the existing fallback to `%Object.prototype%` covering both the non-Object case and a plain bound wrapper whose chain walk yields `undefined`.

To keep `Reflect.construct(Bound, [])` landing on the underlying function's prototype, applied §10.4.1.2 step 5's substitution rule inside the target's bound-unwrap loop: when `SameValue(F, newTarget)` (i.e. default-newTarget calls), substitute `newTarget := F.[[BoundTargetFunction]]` at each level. Explicit non-bound newTargets are preserved.

Added two tests for the edge case the spec demands: a bound newTarget without an own `prototype` falls back to `Object.prototype`, and a bound newTarget with a user-assigned `prototype` uses that value directly.

File: source/units/Goccia.Builtins.GlobalReflect.pas [hunk 2]
```diff
@@ -171,32 +231,71 @@ function TGocciaGlobalReflect.ReflectConstruct(const AArgs: TGocciaArgumentsColl
...
-        // newTarget is a native function — get its .prototype for the instance
-        Result := TGocciaClassValue(Target).Instantiate(CallArgs);
-        ProtoValue := NewTarget.GetProperty(PROP_PROTOTYPE);
+        // newTarget is a non-class constructor — patch the instance prototype
+        // from newTarget.prototype after instantiation (Instantiate only
+        // accepts a class for its newTarget parameter).
+        Result := TGocciaClassValue(EffectiveTarget).Instantiate(CallArgs);
+        ProtoValue := UnwrapBoundFunction(NewTarget).GetProperty(PROP_PROTOTYPE);
         if ProtoValue is TGocciaObjectValue then
           TGocciaObjectValue(Result).Prototype := TGocciaObjectValue(ProtoValue);
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Builtins.GlobalReflect.pas]: Acknowledged on both points; partially addressed in d47f960f.

**Fixed:** the `Object.prototype` fallback. The branch now routes through `GetProtoFromConstructor`, which applies the §10.1.14 fallback to `%Object.prototype%` whenever `newTarget.prototype` is not an Object. The previous `if ProtoValue is TGocciaObjectValue then ... patch` style left the instance with the target class's default prototype on a non-object `newTarget.prototype` — that's gone.

**Not fixed in this PR:** `new.target` propagation through `TGocciaClassValue.Instantiate`. The `Instantiate(args, newTargetClass)` overload only accepts a class for its newTarget parameter, and the constructor call scope doesn't currently bind a separate `new.target` distinct from the active class — that's a pre-existing limitation in the engine, not something this fix introduced. Issue #516 is scoped to making `Reflect.construct` accept function-keyword targets; threading non-class newTarget through class constructors is a larger engine change. Filing as a follow-up rather than expanding scope here.

File: source/units/Goccia.Builtins.GlobalReflect.pas [hunk 3]
```diff
@@ -171,32 +231,71 @@ function TGocciaGlobalReflect.ReflectConstruct(const AArgs: TGocciaArgumentsColl
...
     begin
-      // Target is a native function constructor
-      Result := TGocciaNativeFunctionValue(Target).Call(CallArgs,
+      // Native constructors create their own internal-slot-backed instance;
+      // the receiver is unused so we pass HoleValue. newTarget-driven
+      // prototype propagation for native constructors is not yet wired here.
+      Result := TGocciaNativeFunctionValue(EffectiveTarget).Call(CallArgs,
         TGocciaHoleValue.HoleValue);
-    end;
+    end
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Builtins.GlobalReflect.pas]: Acknowledged but not fixed in this PR.

The `HoleValue` receiver argument to `TGocciaNativeFunctionValue.Call` predates this change — the construction switch's existing native branch already passed it. Threading `newTarget` through every native constructor (Array, TypedArrays, Map, Set, Error subclasses, etc.) so each can run `OrdinaryCreateFromConstructor(newTarget, ...)` is a substantial engine-wide change: the `TGocciaNativeFunctionCallback` signature would need to grow a `newTarget` parameter, and every native constructor would need a spec-matching allocator.

Issue #516 is scoped to the predicate ("function declarations/expressions are rejected") and `isConstructor.js` doesn't probe newTarget propagation through native constructors. Filing this as a follow-up rather than expanding the diff. The new comment block above the native branch already documents the limitation explicitly.

File: source/units/Goccia.Builtins.GlobalReflect.pas (hunk 2 above)
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Builtins.GlobalReflect.pas]: Tracking the `new.target` propagation work as #529.

File: source/units/Goccia.Builtins.GlobalReflect.pas (hunk 3 above)
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Builtins.GlobalReflect.pas]: Tracking the native-constructor newTarget propagation work as #530.

File: source/units/Goccia.Builtins.GlobalReflect.pas [hunk 4]
```diff
@@ -171,32 +222,77 @@ function TGocciaGlobalReflect.ReflectConstruct(const AArgs: TGocciaArgumentsColl
...
+      end;
+      CallArgs.Free;
+      CallArgs := CombinedArgs;
+      EffectiveTarget := BoundFn.OriginalFunction;
+    end;
+
     // Step 4: Return ? Construct(target, args, newTarget)
-    if Target is TGocciaClassValue then
+    if EffectiveTarget is TGocciaProxyValue then
+      Result := TGocciaProxyValue(EffectiveTarget).ConstructTrap(CallArgs)
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Builtins.GlobalReflect.pas]: Good catch — fixed in 321c1fc7.

Widened `TGocciaProxyValue.ConstructTrap` with an optional `ANewTarget` parameter (default `nil` → `Self`, preserving `new proxy(...)` semantics for the existing Evaluator/VM callers) and threaded it through the trap invocation and the chained-proxy fallback. The non-proxy fallback paths still drop newTarget — that's tracked in #529 (classes) and #530 (native constructors).

Added three tests:
- explicit newTarget propagates through the trap
- default newTarget is the proxy itself
- nested-proxy outer fallback forwards newTarget into the inner trap

8633/8633 pass in both interpreter and bytecode modes.

File: source/units/Goccia.Values.ProxyValue.pas [hunk 5]
```diff
@@ -1297,19 +1308,34 @@ function TGocciaProxyValue.ConstructTrap(
...
+    end
     else if FTarget is TGocciaNativeFunctionValue then
       Result := TGocciaNativeFunctionValue(FTarget).Call(AArguments,
         TGocciaHoleValue.HoleValue)
     else if FTarget is TGocciaFunctionBase then
-      Result := TGocciaFunctionBase(FTarget).Call(AArguments,
-        TGocciaHoleValue.HoleValue)
+      Result := ConstructOrdinaryWithReceiver(TGocciaFunctionBase(FTarget),
+        AArguments,
+        TGocciaObjectValue.Create(GetProtoFromConstructor(EffectiveNewTarget)))
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Values.ProxyValue.pas]: Good catch — fixed in d91e4f10.

Extracted a shared `ConstructValue(target, args, newTarget)` helper in `Goccia.Values.FunctionBase.pas` that walks the §10.4.1.2 bound chain (merging bound args, applying the SameValue substitution) before dispatching by target kind. Both `Reflect.construct` and the proxy no-trap fallback now call it, so a `TGocciaBoundFunctionValue` target is unwrapped to its underlying `[[BoundTargetFunction]]` before the synthetic receiver is passed in — `BoundFunction.Call`'s `FBoundThis` forwarding can no longer leak into `new Proxy(F.bind(obj), {})()`.

Added four gated tests covering the affected scenarios:
- proxy fallback over `Capture.bind(sentinel)` — receiver wins over the bound `this`
- proxy fallback over a doubly-bound function — args merged through the chain (sum=6)
- proxy fallback over a bound class — class constructor runs with merged bound args
- the existing function-target tests for prototype propagation through proxy fallback

8641/8641 pass in both interpreter and bytecode modes.

File: source/units/Goccia.Values.ProxyValue.pas [hunk 6]
```diff
@@ -1297,19 +1308,34 @@ function TGocciaProxyValue.ConstructTrap(
...
+      if EffectiveNewTarget is TGocciaClassValue then
+        Result := TGocciaClassValue(FTarget).Instantiate(AArguments,
+          TGocciaClassValue(EffectiveNewTarget))
+      else
+      begin
+        Result := TGocciaClassValue(FTarget).Instantiate(AArguments);
+        if Result is TGocciaObjectValue then
+          TGocciaObjectValue(Result).Prototype :=
+            GetProtoFromConstructor(EffectiveNewTarget);
+      end;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Values.ProxyValue.pas]: Verified — finding is valid but pre-existing, filed as #534 and skipped here.

Probed our engine: `TGocciaClassValue.Instantiate` always returns the freshly-allocated receiver, silently discarding any explicit object return from the constructor body. So today the patch-after-the-fact (mutating `Result.Prototype` after `Instantiate` returns) is sound — there is no user-returned object to corrupt, only the receiver. The `new` operator correctly honors explicit returns through `InstantiateClass` / `InvokeConstructableWithReceiver`, but `Instantiate` itself does not.

Once `Instantiate` is fixed to honor explicit returns (see #534 for the full rationale and reproduction), the prototype-patch in `ConstructValue` must move to a pre-call receiver allocation strategy — either an `Instantiate` overload that accepts a custom receiver/proto, or a delegation to `InvokeConstructableWithReceiver`. Both options are engine-wide changes that overlap heavily with #529, so I have left a comment block in `ConstructValue` flagging the dependency:

```pascal
// Patching the freshly-created instance's [[Prototype]] after the
// constructor returns is sound while Instantiate ignores explicit
// object returns; once that engine limitation is lifted (tracked
// alongside #529), this branch must derive the receiver pre-call.
```

Keeping #516 scoped to the function-keyword predicate and minimum spec-correct dispatch.

File: source/units/Goccia.Builtins.GlobalReflect.pas [hunk 7]
```diff
@@ -158,45 +152,23 @@ function TGocciaGlobalReflect.ReflectConstruct(const AArgs: TGocciaArgumentsColl
...
+  begin
+    NewTarget := AArgs.GetElement(2);
+    if not IsConstructorValue(NewTarget) then
+      ThrowTypeError(SErrorReflectConstructNewTargetMustBeConstructor, SSuggestNotConstructorType);
+  end
+  else
+    NewTarget := Target;
+
   // Step 2: Let args be ? CreateListFromArrayLike(argumentsList)
   CallArgs := CreateListFromArrayLike(ArgsList, 'Reflect.construct');
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Builtins.GlobalReflect.pas:166]: Skipping — finding has the spec order inverted.

ES2026 §28.1.2 [Reflect.construct](https://tc39.es/ecma262/multipage/reflection.html#sec-reflect.construct):

> 1. If IsConstructor(target) is false, throw a TypeError exception.
> 2. If newTarget is not present, set newTarget to target.
> 3. Else if IsConstructor(newTarget) is false, throw a TypeError exception.
> 4. Let args be ? CreateListFromArrayLike(argumentsList).
> 5. Return ? Construct(target, args, newTarget).

Step 3 (newTarget validation) comes **before** step 4 (CreateListFromArrayLike). My current code matches this ordering.

Verified against the reference engine:

```
$ node -e "try { Reflect.construct(class {}, 'not-array', 42); } catch (e) { console.log(e.message); }"
42 is not a constructor
```

V8 throws the newTarget TypeError, not the CreateListFromArrayLike error — confirming step 3 fires before step 4. Test262's existing tests for these two error paths cite the same step numbering (`test/built-ins/Reflect/construct/newtarget-is-not-constructor-throws.js` cites step 3; `arguments-list-is-not-array-like.js` cites step 4).

Re-ordering as suggested would diverge from spec and from V8/SpiderMonkey/JSC.

**Comment by frostney** [comment]
@coderabbitai review

**Comment by frostney** [comment]
@coderabbitai review
